op-batcher/batcher/driver.go

```diff
@@ -945,3 +920,134 @@ func logFields(xs ...any) (fs []any) {
 	}
 	return fs
 }
+
+type ChannelStatuser interface {
+	isFullySubmitted() bool
+	isTimedOut() bool
+	LatestL2() eth.BlockID
+	MaxInclusionBlock() uint64
+}
+
+var SeqOutOfSyncError error = errors.New("sequencer out of sync")
+
+type SyncActions struct {
+	clearState      *eth.BlockID
+	blocksToPrune   int
+	channelsToPrune int
+	blocksToLoad    [2]uint64 // the range [start,end] that should be loaded into the local state.
+	// NOTE this range is inclusive on both ends, which is a change to previous behaviour.
+}
+
+func (s SyncActions) String() string {
+	return fmt.Sprintf(
+		"SyncActions{blocksToPrune: %d, channelsToPrune: %d, clearState: %v, blocksToLoad: [%d, %d]}", s.blocksToPrune, s.channelsToPrune, s.clearState, s.blocksToLoad[0], s.blocksToLoad[1])
+}
+
+// computeSyncActions determines the actions that should be taken based on the inputs provided. The inputs are the current
+// state of the batcher (blocks and channels), the new sync status, and the previous current L1 block. The actions are returned
+// in a struct specifying the number of blocks to prune, the number of channels to prune, whether to wait for node sync, the block
+// range to load into the local state, and whether to clear the state entirely. Returns an error if the sequencer is out of sync.
+func computeSyncActions[T ChannelStatuser](newSyncStatus eth.SyncStatus, prevCurrentL1 eth.L1BlockRef, blocks queue.Queue[*types.Block], channels []T, l log.Logger) (SyncActions, error) {
+
+	if newSyncStatus.HeadL1 == (eth.L1BlockRef{}) {
+		l.Warn("empty sync status")
+		return SyncActions{}, SeqOutOfSyncError
+	}
+
+	if newSyncStatus.CurrentL1.Number < prevCurrentL1.Number {
+		// This can happen when the sequencer restarts
+		l.Warn("sequencer currentL1 reversed")
+		return SyncActions{}, SeqOutOfSyncError
+	}
+
+	oldestBlock, hasBlocks := blocks.Peek()
+
+	if !hasBlocks {
+		s := SyncActions{
+			blocksToLoad: [2]uint64{newSyncStatus.SafeL2.Number + 1, newSyncStatus.UnsafeL2.Number},
+		}
+		l.Info("no blocks in state", "syncActions", s)
+		return s, nil
+	}
+
+	if oldestBlock.NumberU64() > newSyncStatus.SafeL2.Number+1 {
+		s := SyncActions{
+			clearState:   &newSyncStatus.SafeL2.L1Origin,
+			blocksToLoad: [2]uint64{newSyncStatus.SafeL2.Number + 1, newSyncStatus.UnsafeL2.Number},
+		}
```

Comment:
It looks like this exact `SyncAction` is the same in the next 4 or so clear-state cases. So maybe we save it to a `clearSyncAction` and then we can just return the same in the next couple clear-state cases.

Reply:
[organise computeSyncActions and introduce startAfresh syncAction](https://github.com/ethereum-optimism/optimism/pull/13060/commits/8892aa5b4e06fc2ab74859a1eb49be7986d3a9df)